File: tests/chat-binary-codec.test.mjs
import { readFile } from 'node:fs/promises';
import { describe, expect, it } from 'vitest';
import { CHAT_AAD_PREFIX, encodeChatHeader, encodeChatSignatureInput } from '../src/contracts/chat-binary-codec.mjs';

const fixture = JSON.parse(
  await readFile(new URL('./fixtures/protocol/mina-chat-event-v2.json', import.meta.url), 'utf8'),
);
// Les vecteurs vivent dans un fichier LU AUSSI par les tests Kotlin : le contrat
// d'interopérabilité est donc vérifié des deux côtés contre la même source, pas contre deux
// copies qui pourraient diverger.
const vectors = JSON.parse(
  await readFile(new URL('./fixtures/protocol/mina-chat-codec-vectors.json', import.meta.url), 'utf8'),
);

describe('codec binaire canonique du chat', () => {
  it('produit exactement l\'AAD du vecteur partagé avec Kotlin', () => {
    expect(encodeChatHeader(fixture).toString('hex')).toBe(vectors.aadHex);
    expect(encodeChatSignatureInput(fixture).length).toBe(vectors.signatureInputLength);
  });

  it('commence par le préfixe de domaine et la version', () => {
    const header = encodeChatHeader(fixture);
    expect(header.subarray(0, CHAT_AAD_PREFIX.length).toString('ascii')).toBe(CHAT_AAD_PREFIX);
    expect(header.readUInt16BE(CHAT_AAD_PREFIX.length)).toBe(2);
  });

  it('change dès qu\'un SEUL champ du contexte change', () => {
    const base = encodeChatHeader(fixture).toString('hex');
    for (const patch of [
      { threadId: 'thread-autre' },
      { senderDeviceId: 'device-huawei' },
      { deviceSequence: 2 },
      { keyEpoch: 2 },
      { routingClass: 'control' },
      { createdAtMs: fixture.createdAtMs + 1 },
      { expiresAtMs: fixture.expiresAtMs + 1 },
    ]) {
      expect(encodeChatHeader({ ...fixture, ...patch }).toString('hex')).not.toBe(base);
    }
  });

  it('ne confond jamais deux champs voisins (longueurs préfixées)', () => {
    // Sans préfixe de longueur, « ab » + « c » et « a » + « bc » donneraient les mêmes octets.
    const first = encodeChatHeader({ ...fixture, threadId: 'ab', senderDeviceId: 'c' });
    const second = encodeChatHeader({ ...fixture, threadId: 'a', senderDeviceId: 'bc' });
    expect(first.toString('hex')).not.toBe(second.toString('hex'));
  });

  it('l\'entrée de signature couvre l\'en-tête ET le contenu chiffré', () => {
    const base = encodeChatSignatureInput(fixture).toString('hex');
    expect(base.startsWith(Buffer.from('MINA_CHAT_SIGNATURE_V1\0', 'ascii').toString('hex'))).toBe(true);
    // Recoller un autre ciphertext sur le même en-tête doit produire une entrée différente.
    const swapped = encodeChatSignatureInput({ ...fixture, payloadCiphertext: 'YXV0cmU=' });
    expect(swapped.toString('hex')).not.toBe(base);
    const otherNonce = encodeChatSignatureInput({ ...fixture, nonce: 'YWJjZGVmZ2hpams=' });
    expect(otherNonce.toString('hex')).not.toBe(base);
  });

  it('refuse un champ démesuré avant toute allocation', () => {
    expect(() => encodeChatHeader({ ...fixture, threadId: 'x'.repeat(5_000) }))
      .toThrow('chat_codec_champ_trop_long');
  });
});

File: android/core/protocol/src/test/kotlin/fr/mina/gateway/protocol/ChatBinaryCodecTest.kt
package fr.mina.gateway.protocol

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertThrows
import org.junit.Test

/**
 * LE test d'interopérabilité : Kotlin doit produire exactement l'AAD hexadécimal figé dans
 * `mina-chat-codec-vectors.json`, le même fichier que lit le test Node. Si ce test passe des
 * deux côtés, une signature calculée sur PC vérifie sur Android et réciproquement.
 */
class ChatBinaryCodecTest {
    private fun resource(name: String): String {
        val stream = javaClass.classLoader!!.getResourceAsStream(name)
        requireNotNull(stream) { "fixture $name introuvable" }
        return stream.bufferedReader().readText()
    }

    private fun event(): ChatEvent = ChatEventCodec.decode(JSONObject(resource("mina-chat-event-v2.json")))
    private fun vectors(): JSONObject = JSONObject(resource("mina-chat-codec-vectors.json"))

    @Test
    fun `produit exactement l AAD du vecteur partage avec Node`() {
        val expected = vectors().getString("aadHex")
        assertEquals(expected, ChatBinaryCodec.toHex(ChatBinaryCodec.encodeHeader(event())))
    }

    @Test
    fun `produit une entree de signature de la longueur attendue`() {
        val expected = vectors().getInt("signatureInputLength")
        assertEquals(expected, ChatBinaryCodec.encodeSignatureInput(event()).size)
    }

    @Test
    fun `l AAD change des qu un seul champ du contexte change`() {
        val base = ChatBinaryCodec.toHex(ChatBinaryCodec.encodeHeader(event()))
        val patches = listOf(
            event().copy(threadId = "thread-autre"),
            event().copy(senderDeviceId = "device-huawei"),
            event().copy(deviceSequence = 2),
            event().copy(keyEpoch = 2),
            event().copy(routingClass = "control"),
            event().copy(createdAtMs = event().createdAtMs + 1),
            event().copy(expiresAtMs = event().expiresAtMs + 1),
        )
        for (patched in patches) {
            assertNotEquals(base, ChatBinaryCodec.toHex(ChatBinaryCodec.encodeHeader(patched)))
        }
    }

    @Test
    fun `ne confond jamais deux champs voisins`() {
        // Sans préfixe de longueur, « ab » + « c » et « a » + « bc » donneraient les mêmes octets.
        val first = ChatBinaryCodec.encodeHeader(event().copy(threadId = "ab", senderDeviceId = "c"))
        val second = ChatBinaryCodec.encodeHeader(event().copy(threadId = "a", senderDeviceId = "bc"))
        assertNotEquals(ChatBinaryCodec.toHex(first), ChatBinaryCodec.toHex(second))
    }

    @Test
    fun `refuse un champ demesure avant toute allocation`() {
        assertThrows(IllegalArgumentException::class.java) {
            ChatBinaryCodec.encodeHeader(event().copy(threadId = "x".repeat(5_000)))
        }
    }
}
